mongo/mongo.go

```diff
@@ -0,0 +1,189 @@
+		return sess, nil
+	} else if err != mongo.ErrNoDocuments {
+		return nil, errors.Wrap(err, "select")
+	}
+
+	return session.NewBaseSession(sid, s.encoder), nil
+
+}
+
+func (s mongoStore) Destroy(ctx context.Context, sid string) error {
+	_, err := s.db.Collection(s.collection).DeleteOne(ctx, bson.M{"key": sid})
+	if err != nil {
+		return errors.Wrap(err, "delete")
+	}
+	return nil
+}
+
+func (s mongoStore) Save(ctx context.Context, sess session.Session) error {
+	binary, err := sess.Encode()
+	if err != nil {
+		return errors.Wrap(err, "encode")
+	}
+
+	upsert := true
+	_, err = s.db.Collection(s.collection).
+		UpdateOne(ctx, bson.M{"key": sess.ID()}, bson.M{"$set": bson.M{
+			"key":        sess.ID(),
+			"data":       binary,
+			"expired_at": s.nowFunc().Add(s.lifetime).UTC(),
+		}}, &options.UpdateOptions{
+			Upsert: &upsert,
+		})
+	if err != nil {
+		return errors.Wrap(err, "upsert")
+	}
+	return nil
+}
+
+func (s mongoStore) GC(ctx context.Context) error {
+	_, err := s.db.Collection(s.collection).DeleteMany(ctx, bson.M{"expired_at": bson.M{"$lt": s.nowFunc().UTC()}})
+	if err != nil {
+		return errors.Wrap(err, "GC")
+	}
+	return nil
+}
+
+// Options keeps the settings to set up Redis client connection.
+type Options = options.ClientOptions
+
+// Config contains options for the Redis session store.
+type Config struct {
+	// For tests only
+	nowFunc func() time.Time
+	db      *mongo.Database
+
+	// Options is the settings to set up MongoDB client connection.
+	Options *Options
+	// DSN is the database source name to the MongoDB.
+	DSN string
+	// Collection is the collection name for storing session data. Default is "sessions".
+	Collection string
+	// Lifetime is the duration to have no access to a session before being
+	// recycled. Default is 3600 seconds.
+	Lifetime time.Duration
+	// Encoder is the encoder to encode session data. Default is bson.Encoder.
+	Encoder session.Encoder
+	// Decoder is the decoder to decode session data. Default is bson.Decoder.
+	Decoder session.Decoder
+}
+
+// Initer returns the session.Initer for the MongoDB session store.
+func Initer() session.Initer {
+	return func(ctx context.Context, args ...interface{}) (session.Store, error) {
+		var cfg *Config
+		for i := range args {
+			switch v := args[i].(type) {
+			case Config:
+				cfg = &v
+			}
+		}
+
+		if cfg == nil {
+			return nil, fmt.Errorf("config object with the type '%T' not found", Config{})
+		} else if cfg.DSN == "" && cfg.db == nil {
+			return nil, errors.New("empty DSN")
+		}
+
+		if cfg.db == nil {
+			client, err := mongo.Connect(ctx, cfg.Options)
+			if err != nil {
+				return nil, errors.Wrap(err, "open database")
+			}
+			cfg.db = client.Database(cfg.DSN)
+		}
+
+		if cfg.nowFunc == nil {
+			cfg.nowFunc = time.Now
+		}
+		if cfg.Lifetime.Seconds() < 1 {
+			cfg.Lifetime = 3600 * time.Second
+		}
+		if cfg.Collection == "" {
+			cfg.Collection = "sessions"
+		}
+		if cfg.Encoder == nil {
+			cfg.Encoder = session.GobEncoder
```

Comment:
Shouldn't `bson.Encoder` be used here (as said in docstring)?

Reply:
The docstring here is not correct, we should keep using `session.GobEncoder`.